File: 2171/go/main.go
/*
 * @lc app=leetcode.cn id=2171 lang=golang
 *
 * [2171] 拿出最少数目的魔法豆
 */
package main

import "sort"

// @lc code=start
func minimumRemoval(beans []int) int64 {
	length := len(beans)
	if length <= 1 {
		return 0
	}
	sort.Ints(beans)
	var total int64 = 0
	for _, bean := range beans {
		total += int64(bean)
	}
	prevSumArr := make([]int64, length+1)

	var remove int64 = 0
	var result int64 = total

	for i, bean := range beans {
		prevSumArr[i+1] = prevSumArr[i] + int64(bean)
		remove = prevSumArr[i] + (total - prevSumArr[i+1] - int64(length-i-1)*int64(bean))
		result = min(result, remove)
	}
	return result
}
func min(a, b int64) int64 {
	if a < b {
		return a
	}
	return b
}

// @lc code=end

func main() {
	println(minimumRemoval([]int{4, 1, 6, 5}))
	println(minimumRemoval([]int{2, 10, 3, 2}))
	println(minimumRemoval([]int{10}))
	println(minimumRemoval([]int{10, 10}))
}

File: 2171/python3/main.py
#
# @lc app=leetcode.cn id=2171 lang=python3
#
# [2171] 拿出最少数目的魔法豆
#


# @lc code=start
class Solution:
    def minimumRemoval(self, beans: list[int]) -> int:
        length = len(beans)
        if length <= 1:
            return 0
        beans.sort()
        prefix_sum = [0] * length
        total_beans = sum(beans)
        min_removal = float("inf")

        for i in range(length):
            # python 中 prefix_sum[-1] 是最后一位，无须单独处理第一位
            prefix_sum[i] = prefix_sum[i - 1] + beans[i]
            beans_to_remove = prefix_sum[i - 1] + (
                total_beans - prefix_sum[i] - (length - i - 1) * beans[i]
            )
            min_removal = min(min_removal, beans_to_remove)
        return min_removal


# @lc code=end

c = Solution()
print(c.minimumRemoval([4, 1, 6, 5]))
print(c.minimumRemoval([2, 10, 3, 2]))
print(c.minimumRemoval([10]))
print(c.minimumRemoval([10, 10]))
